### PR TITLE
Improve typing/coverage for linter module

### PR DESCRIPTION
**Pull request by pwildenhain**
Related to #488 

Finished adding coverage -- I think this is as good as this one will get (for now). Still a good amount of untested code in `Linter.fix_string` but it's all for fail safes, which we can't anticipate 🤷‍♂️ 

Going to add some type hints before requesting a review.

Side note (@alanmcruickshank)
There was some untested code in the `Linter.fix()` method ~but I didn't address it since it doesn't actually seem like it's used in the CLI. Is this accurate?~ nvm, I can see now that it's used in `Linter.lint_string()`

### CODE REVIEW

File: test/core/linter_test.py [hunk 1]
```diff
@@ -127,3 +139,39 @@ def test__linter__linting_result__combine_dicts():
...
+    all([type(v) == SQLLintError for v in result.get_violations()])
+
+
+def test__linter__raises_malformed_noqa():
+    """A badly formatted noqa gets raised as a parsing error."""
+    lntr = Linter()
+    result = lntr.lint_string_wrapped("select 1 --noqa missing semicolon")
+
+    with pytest.raises(SQLParseError):
+        result.check_tuples()
```
Comment (alanmcruickshank) [inline test/core/linter_test.py:177]: Nice!

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
I'd like to keep the `.tree()` method because I think it's going to be useful for the API. Can we instead get a test case around it?

Even better would be to test the API example (in `examples/`) directly if you can work out a way of doing that nicely?

**Comment by pwildenhain** [comment]
Also, @alanmcruickshank did something happen with codecov.io? I haven't seen to code coverage diff being reported on the last couple PRs

File: src/sqlfluff/core/errors.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
 """Errors - these are closely linked to what used to be called violations."""
-from typing import Optional
+from typing import Optional, Tuple
+
+CheckTuple = Tuple[str, int, int]
```
Comment (dmateusp) [inline src/sqlfluff/core/errors.py:4]: There's type aliases in Python? whaat 🤯

**Review by dmateusp — COMMENTED** [review]
heroic effort 🎉

**Comment by alanmcruickshank** [comment]
> Also, @alanmcruickshank did something happen with codecov.io? I haven't seen to code coverage diff being reported on the last couple PRs

No idea on this - good catch. I'll try removing the plugin and re-adding it.

**Comment by pwildenhain** [comment]
@alanmcruickshank This is good to go I think 🙆‍♂️

**Review by alanmcruickshank — APPROVED** [review]

**Pull request merged**